{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "meth_preprocess_emogi.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "J49FYsY-ygOd",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "941c3993-39d9-4164-d249-d9df15d0ea71"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount(\"/content/drive\")"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import os, h5py, argparse, sys\n",
        "from collections import Counter\n",
        "from joblib import Parallel, delayed"
      ],
      "metadata": {
        "id": "ftd7przUyl91"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def load_annotation_gff(annotation_path, gene_list=None, meth_data=None):\n",
        "    \"\"\"Load a GFF annotation file from disk into a pandas DF.\n",
        "    This function will read a genome annotation file (GFF3 format) from disk\n",
        "    into a dataframe. All annotations that do not originate from genes will\n",
        "    be removed, together with NAs and non-relevant genes (optional, only when\n",
        "    'gene_list' is given).\n",
        "    Parameters:\n",
        "    ----------\n",
        "    annotation_path:            The path to the GFF3 annotation file\n",
        "    gene_list:                  Dataframe with relevant genes as columns.\n",
        "                                Specifies the list of genes to keep the\n",
        "                                annotation for.\n",
        "                                If None, all the annotations are returned.\n",
        "    meth_data:                  Specify a methylation file to use for the\n",
        "                                extraction of promoter regions for each of\n",
        "                                the genes. If None, a simple method is used\n",
        "                                that just takes the 500 bp region around the\n",
        "                                TSS.\n",
        "    Returns:\n",
        "    -------\n",
        "    A pandas DF containing the annotations. All NAs are deleted and only\n",
        "    genes are kept (type column == 'gene'). Furthermore, promoter regions\n",
        "    are already extracted and gene names and IDs are present as distinct\n",
        "    columns.\n",
        "    If a gene_list is given, only annotations for these genes are kept\n",
        "    while all other genes are removed.\n",
        "    \"\"\"\n",
        "    a_df = pd.read_csv(annotation_path, sep='\\t', skiprows=7,\n",
        "                       header=None,\n",
        "                       names=['chr', 'source', 'type', 'start', 'end', 'score',\n",
        "                              'strand', 'phase', 'attr']\n",
        "                       )\n",
        "    # drop all NAs\n",
        "    a_df.dropna(axis=0, inplace=True)\n",
        "    # remove everything that is not a gene\n",
        "    a_df = a_df[a_df.type == 'gene']\n",
        "    # put gene names and ENSEMBL IDs as extra columns\n",
        "    annotated_gene_ids = [i[0].split('=')[1].split('.')[0].strip() for i in a_df.attr.str.split(';')]\n",
        "    annotated_gene_names = [i[3].strip().split('=')[1].strip() for i in a_df.attr.str.split(';')]\n",
        "    a_df['ID'] = annotated_gene_ids\n",
        "    a_df['Symbol'] = annotated_gene_names\n",
        "\n",
        "    # remove duplicate genes (not interested in transcript level)\n",
        "    a_df.drop_duplicates(subset='Symbol', inplace=True)\n",
        "\n",
        "    # extract promoter regions for the genes\n",
        "    if meth_data is None: # use simple 500bp approach\n",
        "        promoter_vec = np.vectorize(get_promotor_window)\n",
        "        p_starts, p_ends = promoter_vec(a_df.start, a_df.end, a_df.strand)\n",
        "    elif 'promoter_start' in a_df.columns and 'promoter_end' in a_df.columns:\n",
        "        p_starts = a_df.promoter_start # don't do anything\n",
        "        p_ends = a_df.promoter_start\n",
        "    else: # use sliding window approach from Lisa\n",
        "        x = a_df.apply(promoter_window_wrapper, axis=1, m_df=meth_data)\n",
        "        p_starts = np.array([i[0] for i in x])\n",
        "        p_ends = np.array([i[1] for i in x])\n",
        "    a_df['promoter_start'] = p_starts\n",
        "    a_df['promoter_end'] = p_ends\n",
        "\n",
        "    # remove all genes that are not in the gene list\n",
        "    if not gene_list is None:\n",
        "        a_df = a_df[a_df.Symbol.isin(gene_list.Symbol)]\n",
        "    return a_df"
      ],
      "metadata": {
        "id": "Cm1ftE433VYX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def load_methylation_file(path):\n",
        "    m_df = pd.read_csv(path, sep='\\t')\n",
        "    m_df.dropna(axis=0, inplace=True)\n",
        "    return m_df\n",
        "\n",
        "\n",
        "def load_relevant_genes(container_path):\n",
        "    with h5py.File(container_path, 'r') as f:\n",
        "        node_names = f['gene_names'][:]\n",
        "    # build data frame for node names and return\n",
        "    return pd.DataFrame(node_names, columns=['ID', 'Symbol'])"
      ],
      "metadata": {
        "id": "0Mt-GD-i3b4X"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def promoter_window_wrapper(row, m_df):\n",
        "    return get_promotor_window(row['start'], row['end'], row['strand'], m_df)\n",
        "\n",
        "\n",
        "def get_promotor_window(start, end, strand, meth_data=None):\n",
        "    if strand == '+':\n",
        "        tss = start\n",
        "    else:\n",
        "        tss = end\n",
        "    if not meth_data is None: # fancy method to get promoter\n",
        "        scan_region = (np.max(tss-1000, 0), tss + 1000)\n",
        "        return calculate_promoter_window(scan_region[0], scan_region[1], meth_data)\n",
        "    else:\n",
        "        return np.max(tss-250, 0), tss + 250\n",
        "\n",
        "\n",
        "def calculate_promoter_window(scan_start, scan_end, meth_data, shift=50, size=200):\n",
        "    best_mean_meth = None\n",
        "    best_window = (None, None)\n",
        "    for i in range(int(scan_start), int(scan_end-size+1), shift):\n",
        "        # get mean meth for window\n",
        "        m_sites_window = meth_data[meth_data.Start.between(i, i+size)]\n",
        "        m_mean_window = m_sites_window.Beta_value_mean.mean()\n",
        "        \n",
        "        # decide on best window\n",
        "        if best_mean_meth is None: # first window\n",
        "            best_mean_meth = m_mean_window\n",
        "            best_window = i, i+size\n",
        "        else: # any of the later windows\n",
        "            if abs(best_mean_meth-m_mean_window) > 0.25: # large change in mean detected\n",
        "                break\n",
        "            else: # make the current window the best one\n",
        "                best_mean_meth = m_mean_window\n",
        "                best_window = i, i+size\n",
        "    return best_window"
      ],
      "metadata": {
        "id": "KJL85pQm3lKZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_filenames(meth_raw_dir):\n",
        "    \"\"\"Extract all valid methylation files in dir.\n",
        "    \"\"\"\n",
        "    all_filenames = []\n",
        "    skipped_files = 0\n",
        "    for dname in os.listdir(meth_raw_dir):\n",
        "        sub_dirname = os.path.join(meth_raw_dir, dname)\n",
        "        if os.path.isdir(sub_dirname):\n",
        "            for fname in os.listdir(sub_dirname):\n",
        "                if fname.endswith('gdc_hg38.txt'):\n",
        "                    # don't use 27k methylation arrays\n",
        "                    if fname.split('.')[2].endswith('450'):\n",
        "                        all_filenames.append(os.path.join(sub_dirname, fname))\n",
        "                    else:\n",
        "                        skipped_files += 1\n",
        "    print (\"Skipped {} 27k arrays\".format(skipped_files))\n",
        "    return all_filenames"
      ],
      "metadata": {
        "id": "MmhPt3lW3sQn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_mean_betaval_for_sample(annotation_genes, methylation_levels):\n",
        "    \"\"\"Extract mean methylation levels in promoters/gene bodies for each annotated gene.\n",
        "    \n",
        "    This function looks for all the methylation sites that fall\n",
        "    inside all the promoters and gene bodies in the annotation genes df.\n",
        "    The annotation_genes df is expected to contain the columns\n",
        "    'promoter_start' and 'promoter_end'.\n",
        "    \n",
        "    Parameters:\n",
        "    ----------\n",
        "    annotation_genes:           A dataframe containing a gene per row.\n",
        "                                It needs to have columns for promoter_start\n",
        "                                and promoter_end.\n",
        "    methylation_levels:         A dataframe that contains cpg sites per row.\n",
        "                                The sites are expected to be one bp long and\n",
        "                                should have the columns: 'Start', 'End' and \n",
        "                                Beta_value'.\n",
        "    Returns:\n",
        "    Two lists with the mean beta values for each promoter and the number of cpg sites\n",
        "    that support the promoter. The first list contains the mean beta values and\n",
        "    the second list contains the support.\n",
        "    \"\"\"\n",
        "    beta_vals_prom = []\n",
        "    beta_vals_gene = []\n",
        "    n_supports_prom = []\n",
        "    n_supports_gene = []\n",
        "    count = 0\n",
        "    for _, row in annotation_genes.iterrows():\n",
        "        # restrict possible sites to same chr\n",
        "        cpgs_on_chr = methylation_levels[methylation_levels.Chromosome == row.chr]\n",
        "        # promoter\n",
        "        m_sites_in_promoter = cpgs_on_chr[cpgs_on_chr.Start.between(row.promoter_start, row.promoter_end)]\n",
        "        beta_vals_prom.append(m_sites_in_promoter.Beta_value.mean())\n",
        "        n_supports_prom.append(m_sites_in_promoter.shape[0])\n",
        "        # gene body. Don't start with the start and end coordinates but rather with the end of the promoter\n",
        "        if row.strand == '+':\n",
        "            m_sites_gene = cpgs_on_chr[cpgs_on_chr.Start.between(row.promoter_end, row.end)]\n",
        "        else:\n",
        "            m_sites_gene = cpgs_on_chr[cpgs_on_chr.Start.between(row.start, row.promoter_start)]\n",
        "        beta_vals_gene.append(m_sites_gene.Beta_value.mean())\n",
        "        n_supports_gene.append(m_sites_gene.shape[0])\n",
        "        count += 1\n",
        "    return beta_vals_prom, beta_vals_gene, n_supports_prom, n_supports_gene\n"
      ],
      "metadata": {
        "id": "x14qzAW-3yvs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_float(string):\n",
        "    \"\"\" Convert a string into a float.\n",
        "    This function converts a string to float if the string represents a floating\n",
        "    point number (or integer). If that is not the case, it returns NaN.\n",
        "    Parameters:\n",
        "    ----------\n",
        "    string:                     The string that is to be converted to float.\n",
        "    Returns:\n",
        "    A floating point number. Either with the value from `string` or numpy.nan\n",
        "    \"\"\"\n",
        "    try:\n",
        "        return float(string)\n",
        "    except:\n",
        "        return np.inf"
      ],
      "metadata": {
        "id": "f8g3w3Da32hD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_closest_gene(row):\n",
        "    \"\"\" Get the closest gene from a row of TCGA level 3 methylation data.\n",
        "    Extracts the gene name of the closest transcript promoter from\n",
        "    a row of TCGA DNA methylation data (level 3). Each of the lines\n",
        "    contain the coordinates of the CpG site together with information\n",
        "    on the closest gene around it. There are three relevant fields (columns)\n",
        "    for that: Gene_Symbol, Transcript_ID and Position_to_TSS.\n",
        "    \n",
        "    Parameters:\n",
        "    ----------\n",
        "    row:                        A row from a TCGA DNA methylation Dataframe.\n",
        "                                Should contain the columns `Position_to_TSS`\n",
        "                                and `Gene_Symbol`.\n",
        "    \"\"\"\n",
        "    # extract rows\n",
        "    genes = np.array(row.Gene_Symbol.split(';'))\n",
        "    protein_coding_genes = np.array([i == 'protein_coding' for i in row.Gene_Type.split(';')])\n",
        "    dists = np.array([get_float(i) for i in row.Position_to_TSS.split(';')])\n",
        "\n",
        "    # remove non-protein-coding genes\n",
        "    dists = dists[protein_coding_genes]\n",
        "    genes = genes[protein_coding_genes]\n",
        "\n",
        "    # return closest gene and distance\n",
        "    if dists.shape[0] > 0:\n",
        "        idx = np.argmin(np.abs(dists))\n",
        "        return genes[idx], dists[idx]\n",
        "    else:\n",
        "        return None, None"
      ],
      "metadata": {
        "id": "SBB9O-AD37VY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_closest_5prime_transcript(row, gene_transcript_mapping):\n",
        "    \"\"\" Get the closest 5' transcript from a row of TCGA level 3 methylation data.\n",
        "    Extracts the transcript ID of the transcript whose TSS is closest to the\n",
        "    CpG site in question. This is computed from a row of TCGA DNA methylation\n",
        "    data (level 3). The function uses a pre-computed mapping from genes\n",
        "    to 5' transcripts for its calculation.\n",
        "    It first extracts the transcripts and distances to TSS for the CpG site\n",
        "    and then discards all that are not in the list of closest transcripts in\n",
        "    the mapping. Finally, from the valid transcripts the one with the closest\n",
        "    TSS is chosen and returned.\n",
        "    \n",
        "    Parameters:\n",
        "    ----------\n",
        "    row:                        A row from a TCGA DNA methylation Dataframe.\n",
        "                                Should contain the columns `Position_to_TSS`\n",
        "                                and `Gene_Symbol`.\n",
        "    \n",
        "    Returns:\n",
        "    The transcript ID and distance to the transcript TSS of a CpG site.\n",
        "    This can be used as part of an apply call on a dataframe.\n",
        "    \"\"\"\n",
        "    # extract rows\n",
        "    transcripts = pd.Series(row.Transcript_ID.split(';'))\n",
        "    transcripts = transcripts.str.split('.').str[0] # remove transcript version number\n",
        "    \n",
        "    valid_transcripts = transcripts.isin(gene_transcript_mapping)\n",
        "    dists = np.array([get_float(i) for i in row.Position_to_TSS.split(';')])\n",
        "\n",
        "    # remove non-protein-coding genes\n",
        "    dists = dists[valid_transcripts]\n",
        "    transcripts = transcripts[valid_transcripts]\n",
        "\n",
        "    if dists.shape[0] > 0:\n",
        "        idx = np.argmin(np.abs(dists))\n",
        "        return np.array(transcripts)[idx], dists[idx]\n",
        "    else:\n",
        "        return None, None"
      ],
      "metadata": {
        "id": "2pRqPdrA3_uO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_promoter_betaval_tcgaannotation(methylation_levels):\n",
        "    # write to DF which is the closest gene\n",
        "    x = methylation_levels.apply(get_closest_gene, axis=1)\n",
        "    methylation_levels['closest_gene'] = [i[0] for i in x]\n",
        "    methylation_levels['dist_closest_gene'] = [i[1] for i in x]\n",
        "\n",
        "    # filter\n",
        "    close_cpgs = methylation_levels[methylation_levels.dist_closest_gene.abs() < 1000]\n",
        "\n",
        "    # group and return\n",
        "    cpgs_for_genes = close_cpgs.groupby('closest_gene')\n",
        "    beta_values_prom = cpgs_for_genes.Beta_value.mean()\n",
        "    n_supports_prom = cpgs_for_genes.Beta_value.count()\n",
        "\n",
        "    return beta_values_prom, n_supports_prom\n"
      ],
      "metadata": {
        "id": "dK8jSlZI4G-H"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_gene2transcript_map(annotation_df):\n",
        "    \"\"\"Calculate a mapping from a gene to its most 5' transcript.\n",
        "    This function calculates a mapping from a gene to its most 5' transcript using a\n",
        "    GENCODE annotation file. For each transcript, the corresponding parent gene is\n",
        "    extracted together with its distance to the chromosome start. For genes on the\n",
        "    - strand, negative values are used.\n",
        "    Using a map/reduce pattern, the transcripts are grouped by parent genes and the\n",
        "    most 5' transcript per gene is chosen.\n",
        "    Parameters:\n",
        "    ----------\n",
        "    annotation_df:              The GENCODE annotation file as pandas DataFrame.\n",
        "                                Usually, the basic annotation is preferred over the\n",
        "                                comprehensive.\n",
        "    Returns:\n",
        "    A DataFrame containing the gene symbol as index and the corresponding 5' transcript\n",
        "    as column.\n",
        "    \"\"\"\n",
        "    # get a mapping from gene to transcript\n",
        "    transcripts = annotation_df[annotation_df.type == 'transcript'].copy()\n",
        "    # extract all attributes of transcripts\n",
        "    transcript_attrs = []\n",
        "    for attrs in transcripts.attr.str.split(';'):\n",
        "        transcript_attrs.append({i.split('=')[0]:i.split('=')[1] for i in attrs})\n",
        "\n",
        "    # set the attributes to additional columns (remove transcript version number)\n",
        "    transcripts['ID'] = [i['ID'].strip().split('.')[0] for i in transcript_attrs]\n",
        "    transcripts['Parent'] = [i['Parent'] for i in transcript_attrs]\n",
        "    transcripts['Gene_Symbol'] = [i['gene_name'] for i in transcript_attrs]\n",
        "\n",
        "    # build column with the distance to the chromosome start (for transcripts on the - strand, take negative values)\n",
        "    transcripts['dist_to_chr_start'] = transcripts.start\n",
        "    transcripts.loc[transcripts.strand == '-', 'dist_to_chr_start'] = -transcripts.end\n",
        "    transcripts.set_index('ID', inplace=True)\n",
        "    # mapping is now simply the transcript in a group of a gene that has minimum distance\n",
        "    gene_transcript_mapping = transcripts.groupby('Gene_Symbol').dist_to_chr_start.idxmin()\n",
        "\n",
        "    return gene_transcript_mapping"
      ],
      "metadata": {
        "id": "WbHbLOI74y81"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_cpg_transcript_map(annotation_df, methylation_levels):\n",
        "    \"\"\"Calculates a mapping from CpG site to closest transcript.\n",
        "    \n",
        "    This function calculates a mapping from CpG site to gene promoter and\n",
        "    the distance to it. It makes use of the annotations in a TCGA methylation\n",
        "    file (level 3). For each CpG site, its closest transcripts and their distance\n",
        "    to the TSS are extracted. Then, all transcripts that are not the most 5'\n",
        "    transcripts of a gene are discarded and from the remaining transcripts, the\n",
        "    closest one is chosen. This implies that a CpG site can only belong to one\n",
        "    gene promoter.\n",
        "    The calculated map of CpG sites to transcript promoters is finally joined\n",
        "    with a mapping from transcripts to genes (computed from GENCODE annotation)\n",
        "    and the final mapping is returned.\n",
        "    Parameters:\n",
        "    ----------\n",
        "    annotation_df:              The GENCODE annotation file as pandas DataFrame.\n",
        "                                Usually, the basic annotation is preferred over the\n",
        "                                comprehensive.\n",
        "    methylation_levels:         A TCGA level 3 methylation file. This will\n",
        "                                serve as template for all other samples, so\n",
        "                                if you use 450k data, this file should contain\n",
        "                                all the CpG sites you are interested in.\n",
        "    Returns:\n",
        "    A Dataframe with CpG sites as rows (Composite Element REF) and the gene/transcript\n",
        "    as columns. The columns contain the distance to the closest transcript of that CpG\n",
        "    site, the Ensembl ID of that transcript and the gene name of the transcript (Symbol).\n",
        "    \"\"\"\n",
        "    gene_transcript_map = get_gene2transcript_map(annotation_df)\n",
        "    # get the closest 5' transcript according to the gene transcript map\n",
        "    x = methylation_levels.apply(get_closest_5prime_transcript,\n",
        "                                 axis=1,\n",
        "                                 gene_transcript_mapping=gene_transcript_map)\n",
        "    methylation_levels['closest_transcript'] = [i[0] for i in x]\n",
        "    methylation_levels['dist_closest_transcript'] = [i[1] for i in x]\n",
        "    \n",
        "    # get transcript ID and Gene Symbol in mapping as new columns\n",
        "    gene_transcript_map_df = pd.DataFrame(gene_transcript_map)\n",
        "    gene_transcript_map_df.columns = ['Transcript_ID']\n",
        "    gene_transcript_map_df['Symbol'] = gene_transcript_map_df.index\n",
        "\n",
        "    # construct the mapping\n",
        "    cpgs_with_genes = methylation_levels[~methylation_levels.closest_transcript.isnull()]\n",
        "    cpgs_with_genes.drop(['Transcript_ID', 'Position_to_TSS'], axis=1, inplace=True)\n",
        "    cpg_gene_mapping = cpgs_with_genes.merge(gene_transcript_map_df,\n",
        "                                             left_on='closest_transcript',\n",
        "                                             right_on='Transcript_ID'\n",
        "                                            )\n",
        "    cpg_gene_mapping.set_index('Composite Element REF', inplace=True)\n",
        "    mapping_cols = ['dist_closest_transcript', 'Transcript_ID', 'Symbol']\n",
        "    return cpg_gene_mapping[mapping_cols]"
      ],
      "metadata": {
        "id": "SBooaccnEJAR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_meth_df_from_mapping(cpg_gene_map, path_name, clean=False):\n",
        "    \"\"\"Computes promoter DNA methylation from a CpG to gene mapping.\n",
        "    This function computes the average DNA methylation at promoters from\n",
        "    a mapping of CpG sites to genes. It essentially only joins the mapping\n",
        "    with the actual beta values from the sample in question.\n",
        "    It then aggregates the genes together, returning a DataFrame with\n",
        "    the average DNA methylation at the promoter of each gene and the\n",
        "    cancer type of the individual.\n",
        "    Parameters:\n",
        "    ----------\n",
        "    cpg_gene_map:               A mapping between CpG sites and genes/transcripts.\n",
        "                                Can be computed from TCGA annotations to the methylation\n",
        "                                data. @see get_cpg_transcript_map\n",
        "    path_name:                  The path to the methylation file. Should be in TCGA format,\n",
        "                                thus containing information on close-by genes additionally\n",
        "                                to the pure coordinates and beta value.\n",
        "    clean:                      Whether to use potential previous runs and data written into\n",
        "                                the download directory or not. If changing the method of how\n",
        "                                DNA methylation is computed, clean=True is recommended.\n",
        "    Returns:\n",
        "    A dataframe with methylation levels at promoters and number of CpG sites per gene promoter as\n",
        "    columns and gene names as rows and the cancer type of the sample.\n",
        "    Ideally, there should be as many genes as there are in the mapping file.\n",
        "    The columns of the returned dataframe will have the format:\n",
        "    <sample_id>|<cancer_type>|<beta_val/support> (separator is |)\n",
        "    \"\"\"\n",
        "    sub_dirname = os.path.dirname(path_name)\n",
        "    fname = os.path.basename(path_name)\n",
        "    avg_meth_result_path = os.path.join(sub_dirname,\n",
        "            'avg_methylation_tcgaannot.tsv'\n",
        "            )\n",
        "    cancer_type = fname.split('.')[1].split('_')[1].strip().lower()\n",
        "    sample_id = fname.split('.')[-3].strip()\n",
        "\n",
        "    if not os.path.isfile(avg_meth_result_path) or clean:\n",
        "        meth_df = load_methylation_file(path_name)\n",
        "        meth_df.drop(['Transcript_ID', 'Gene_Symbol', 'Gene_Type', 'Position_to_TSS',\n",
        "                      'CGI_Coordinate', 'Feature_Type'],\n",
        "                      axis=1, inplace=True)\n",
        "        meth_df.set_index('Composite Element REF', inplace=True)\n",
        "        meth_annot = meth_df.join(cpg_gene_map)\n",
        "        meth_annot_inrange = meth_annot[meth_annot.dist_closest_transcript.abs() < 1000]\n",
        "        res = pd.DataFrame(meth_annot_inrange.dropna().groupby('Symbol').Beta_value.mean())\n",
        "        res['support_promoter'] = meth_annot_inrange.dropna().groupby('Symbol').Beta_value.count()\n",
        "        colnames = ['mean_beta_value_promoter', 'support_promoter']\n",
        "        cols = ['{}|{}|{}'.format(sample_id, cancer_type, i) for i in colnames]\n",
        "        res.columns = cols\n",
        "        res.to_csv(avg_meth_result_path, sep='\\t')\n",
        "    else:\n",
        "        res = pd.read_csv(avg_meth_result_path, sep='\\t')\n",
        "        res.set_index('Symbol', inplace=True)\n",
        "    return res, cancer_type"
      ],
      "metadata": {
        "id": "4B51yu5L521Q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_meth_df_for_sample(annotation_df, path_name, clean=False, tcga_annot=True):\n",
        "    \"\"\"Calculates the mean methylation DF for one sample.\n",
        "    \n",
        "    This function computes the mean methylation level at\n",
        "    the promoter and gene body for all genes in the annotation\n",
        "    dataframe using the methylation data specified in the path\n",
        "    name. Note that this must be in the TCGA methylation format.\n",
        "    The annotation dataframe is expected to contain the promoter\n",
        "    region in the columns 'promoter_start' and 'promoter_end'.\n",
        "    If the directory of the file given by 'path_name' contains\n",
        "    already a file 'avg_methylation.tsv', this will be loaded\n",
        "    and returned simply instead of re-extracting the methylation\n",
        "    values. If clean is set to True, the extraction takes place\n",
        "    in any case.\n",
        "    Parameters:\n",
        "    ----------\n",
        "    annotation_df:              The annotation file as dataframe. Can be computed using\n",
        "                                @see load_annotation_gff from the gencode annotation.\n",
        "    path_name:                  The path to the methylation file. Should be in TCGA format,\n",
        "                                thus containing information on close-by genes additionally\n",
        "                                to the pure coordinates and beta value.\n",
        "    clean:                      Whether to use potential previous runs and data written into\n",
        "                                the download directory or not. If changing the method of how\n",
        "                                DNA methylation is computed, clean=True is recommended.\n",
        "    tcga_annot:                 Whether or not to use the annotation from TCGA to each\n",
        "                                individual CpG site. Otherwise it uses promoter windows\n",
        "                                around the TSS of a gene.\n",
        "    Returns:\n",
        "    A dataframe with methylation levels at promoters and number of CpG sites per gene promoter as\n",
        "    columns and gene names as rows and the cancer type of the sample.\n",
        "    Ideally, there should be as many genes as there are in the mapping file.\n",
        "    The columns of the returned dataframe will have the format:\n",
        "    <sample_id>|<cancer_type>|<beta_val/support> (separator is |)\n",
        "    \"\"\"\n",
        "    sub_dirname = os.path.dirname(path_name)\n",
        "    fname = os.path.basename(path_name)\n",
        "    if tcga_annot:\n",
        "        avg_meth_result_path = os.path.join(sub_dirname,\n",
        "                'avg_methylation_tcgaannot.tsv'\n",
        "                )\n",
        "    else:\n",
        "        avg_meth_result_path = os.path.join(sub_dirname,\n",
        "                        'avg_methylation.tsv'\n",
        "                        )\n",
        "    # get the cancer type first\n",
        "    cancer_type = fname.split('.')[1].split('_')[1].strip().lower()\n",
        "    sample_id = fname.split('.')[-3].strip()\n",
        "    if not os.path.isfile(avg_meth_result_path) or clean:\n",
        "        # get mean methylation levels around promoters\n",
        "        meth_df = load_methylation_file(path_name)\n",
        "        if tcga_annot:\n",
        "            beta_prom, sup_prom = get_promoter_betaval_tcgaannotation(meth_df)\n",
        "            res = pd.DataFrame([beta_prom, sup_prom]).T\n",
        "            colnames = ['mean_beta_value_promoter', 'support_promoter']\n",
        "            cols = ['{}|{}|{}'.format(sample_id, cancer_type, i) for i in colnames]\n",
        "            res.columns = cols\n",
        "        else:\n",
        "            beta_p, beta_g, sup_p, sup_g = get_mean_betaval_for_sample(annotation_df,\n",
        "                                                                       meth_df)\n",
        "            # construct dataframe from results\n",
        "            res = pd.DataFrame([beta_p, sup_p, beta_g, sup_g]).T\n",
        "            colnames = ['mean_beta_value_promoter', 'support_promoter',\n",
        "                        'mean_beta_value_gene', 'support_gene']\n",
        "            cols = ['{}|{}|{}'.format(sample_id, cancer_type, i) for i in colnames]\n",
        "            res.columns = cols\n",
        "            res.set_index(annotation_df.Symbol, inplace=True) # assumes same order\n",
        "\n",
        "        # write the average beta values (at promoters) to file\n",
        "        res.to_csv(avg_meth_result_path, sep='\\t')\n",
        "    else:\n",
        "        res = pd.read_csv(avg_meth_result_path, sep='\\t')\n",
        "        res.set_index('Symbol', inplace=True)\n",
        "    return res, cancer_type"
      ],
      "metadata": {
        "id": "-ArfHAu659Iu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# main\n",
        "#load data and compute promoter windows if required\n",
        "# if not args.rel_genes is None:\n",
        "#     relevant_genes = load_relevant_genes(args.rel_genes)\n",
        "# else:\n",
        "relevant_genes = None\n",
        "# if args.annotation.endswith('.tsv'):\n",
        "#     annotation_df = pd.read_csv(args.annotation, sep='\\t')\n",
        "# elif args.annotation.endswith('.gff3'):\n",
        "annotation_df = load_annotation_gff('/content/drive/MyDrive/BoltBio/gencode.v39.annotation.gff3')\n",
        "# else:\n",
        "#     print (\"Unknown Annotation file format. Only tsv and gff3 are supported\")\n",
        "#     sys.exit(-1)\n",
        "print (\"Loaded Annotation file with {} genes\".format(annotation_df.shape[0]))\n",
        "\n",
        "# get file names of DNA methylation downloads\n",
        "all_files = get_filenames('/content/drive/MyDrive/BoltBio')\n",
        "print (\"Found {} methylation profiles\".format(len(all_files)))\n",
        "\n",
        "# Derive DNA methylation at promoters\n",
        "# if args.tcga_annot and args.fiveprime: # use TCGA annotations and only fiveprime transcript\n",
        "#     # get mapping of CpG site to closest 5' transcript\n",
        "#     if args.annotation.endswith('.gff3'):\n",
        "a_df = pd.read_csv('/content/drive/MyDrive/BoltBio/gencode.v39.annotation.gff3', sep='\\t', skiprows=7,\n",
        "                        header=None,\n",
        "                        names=['chr', 'source', 'type', 'start', 'end', 'score',\n",
        "                                'strand', 'phase', 'attr'])\n",
        "    # else:\n",
        "    #     a_df = pd.read_csv(args.annotation, sep='\\t')\n",
        "    # cpg_genemap_file = os.path.join(os.path.dirname(args.annotation), 'cpg2genemap.tsv')\n",
        "    # if not os.path.exists(cpg_genemap_file):\n",
        "print (\"Computing CpG site to transcript map\")\n",
        "cpg_mapping = get_cpg_transcript_map(a_df, load_methylation_file(all_files[0]))\n",
        "cpg_mapping.to_csv('/content/drive/MyDrive/BoltBio/cpg_genemap_file', sep='\\t')"
      ],
      "metadata": {
        "id": "9DXEqLQv6CT2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 399
        },
        "outputId": "4f5a3162-0302-4d15-edb9-92de9a693ce0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Loaded Annotation file with 60286 genes\n",
            "Skipped 0 27k arrays\n",
            "Found 2 methylation profiles\n",
            "Computing CpG site to transcript map\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  errors=errors,\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-17-bdd1b880f65b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"Computing CpG site to transcript map\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0mcpg_mapping\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_cpg_transcript_map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mload_methylation_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_files\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 34\u001b[0;31m \u001b[0mcpg_mapping\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcpg_genemap_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     35\u001b[0m \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"Mapping computed. Applying for all samples now...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[0;31m# else:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'cpg_genemap_file' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# print (\"Mapping computed. Applying for all samples now...\")\n",
        "    # else:\n",
        "    #     print (\"Loading mapping of CpG site to transcripts from disk...\")\n",
        "    #     cpg_mapping = pd.read_csv(cpg_genemap_file, sep='\\t')\n",
        "    #     cpg_mapping.set_index('Composite Element REF', inplace=True)\n",
        "    \n",
        "    # use the mapping as template for all samples (e.g. apply)\n",
        "print (\"Applying cpg 2 transcript mapping for all samples (takes time and uses processors)\")\n",
        "results = Parallel(n_jobs=2)(delayed(get_meth_df_from_mapping)(cpg_mapping,\n",
        "                                                                              f,\n",
        "                                                                              False) for f in all_files)\n",
        "# else: # use TCGA annotations (and mean over all transcripts of each gene) or fixed promoters\n",
        "#     results = Parallel(n_jobs=args.n_jobs)(delayed(get_meth_df_for_sample)(annotation_df,\n",
        "#                                                                             f,\n",
        "#                                                                             args.clean,\n",
        "#                                                                             args.tcga_annot) for f in all_files)\n",
        "all_samples_preprocessed = [i[0] for i in results]\n",
        "cancer_types = [i[1] for i in results]\n",
        "\n",
        "# some information on how many samples we have\n",
        "for key, value in Counter(cancer_types).items():\n",
        "    print (\"Found {} samples for Cancer {}\".format(value, key))\n",
        "\n",
        "# join samples to form one df with all samples as columns\n",
        "total_df = pd.concat(all_samples_preprocessed, axis=1)\n",
        "total_df.to_csv('/content/drive/MyDrive/BoltBio/mean_meth_matrix.tsv', sep='\\t')\n",
        "print (\"Finished. Written mean methylation matrix for {} samples to disk\".format(len(all_samples_preprocessed)))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8f9h2E7JgVuy",
        "outputId": "8706fcf5-281a-418e-dc22-5002bc317bb8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Applying cpg 2 transcript mapping for all samples (takes time and uses processors)\n",
            "Found 2 samples for Cancer luad\n",
            "Finished. Written mean methylation matrix for 2 samples to disk\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "OizHPELAi4uT"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}