{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "ftd7przUyl91"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, h5py, argparse, sys\n",
    "from collections import Counter\n",
    "from joblib import Parallel, delayed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Cm1ftE433VYX"
   },
   "outputs": [],
   "source": [
    "def load_annotation_gff(annotation_path, gene_list=None, meth_data=None):\n",
    "    \"\"\"Load a GFF annotation file from disk into a pandas DF.\n",
    "    This function will read a genome annotation file (GFF3 format) from disk\n",
    "    into a dataframe. All annotations that do not originate from genes will\n",
    "    be removed, together with NAs and non-relevant genes (optional, only when\n",
    "    'gene_list' is given).\n",
    "    Parameters:\n",
    "    ----------\n",
    "    annotation_path:            The path to the GFF3 annotation file\n",
    "    gene_list:                  Dataframe with relevant genes as columns.\n",
    "                                Specifies the list of genes to keep the\n",
    "                                annotation for.\n",
    "                                If None, all the annotations are returned.\n",
    "    meth_data:                  Specify a methylation file to use for the\n",
    "                                extraction of promoter regions for each of\n",
    "                                the genes. If None, a simple method is used\n",
    "                                that just takes the 500 bp region around the\n",
    "                                TSS.\n",
    "    Returns:\n",
    "    -------\n",
    "    A pandas DF containing the annotations. All NAs are deleted and only\n",
    "    genes are kept (type column == 'gene'). Furthermore, promoter regions\n",
    "    are already extracted and gene names and IDs are present as distinct\n",
    "    columns.\n",
    "    If a gene_list is given, only annotations for these genes are kept\n",
    "    while all other genes are removed.\n",
    "    \"\"\"\n",
    "    a_df = pd.read_csv(annotation_path, sep='\\t', skiprows=7,\n",
    "                       header=None,\n",
    "                       names=['chr', 'source', 'type', 'start', 'end', 'score',\n",
    "                              'strand', 'phase', 'attr']\n",
    "                       )\n",
    "    # drop all NAs\n",
    "    a_df.dropna(axis=0, inplace=True)\n",
    "    # remove everything that is not a gene\n",
    "    a_df = a_df[a_df.type == 'gene']\n",
    "    # put gene names and ENSEMBL IDs as extra columns\n",
    "    annotated_gene_ids = [i[0].split('=')[1].split('.')[0].strip() for i in a_df.attr.str.split(';')]\n",
    "    annotated_gene_names = [i[3].strip().split('=')[1].strip() for i in a_df.attr.str.split(';')]\n",
    "    a_df['ID'] = annotated_gene_ids\n",
    "    a_df['Symbol'] = annotated_gene_names\n",
    "\n",
    "    # remove duplicate genes (not interested in transcript level)\n",
    "    a_df.drop_duplicates(subset='Symbol', inplace=True)\n",
    "\n",
    "    # extract promoter regions for the genes\n",
    "    if meth_data is None: # use simple 500bp approach\n",
    "        promoter_vec = np.vectorize(get_promotor_window)\n",
    "        p_starts, p_ends = promoter_vec(a_df.start, a_df.end, a_df.strand)\n",
    "    elif 'promoter_start' in a_df.columns and 'promoter_end' in a_df.columns:\n",
    "        p_starts = a_df.promoter_start # don't do anything\n",
    "        p_ends = a_df.promoter_start\n",
    "    else: # use sliding window approach from Lisa\n",
    "        x = a_df.apply(promoter_window_wrapper, axis=1, m_df=meth_data)\n",
    "        p_starts = np.array([i[0] for i in x])\n",
    "        p_ends = np.array([i[1] for i in x])\n",
    "    a_df['promoter_start'] = p_starts\n",
    "    a_df['promoter_end'] = p_ends\n",
    "\n",
    "    # remove all genes that are not in the gene list\n",
    "    if not gene_list is None:\n",
    "        a_df = a_df[a_df.Symbol.isin(gene_list.Symbol)]\n",
    "    return a_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "0Mt-GD-i3b4X"
   },
   "outputs": [],
   "source": [
    "def load_methylation_file(path):\n",
    "    m_df = pd.read_csv(path, sep='\\t')\n",
    "    m_df.dropna(axis=0, inplace=True)\n",
    "    return m_df\n",
    "\n",
    "\n",
    "def load_relevant_genes(container_path):\n",
    "    with h5py.File(container_path, 'r') as f:\n",
    "        node_names = f['gene_names'][:]\n",
    "    # build data frame for node names and return\n",
    "    return pd.DataFrame(node_names, columns=['ID', 'Symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "KJL85pQm3lKZ"
   },
   "outputs": [],
   "source": [
    "def promoter_window_wrapper(row, m_df):\n",
    "    return get_promotor_window(row['start'], row['end'], row['strand'], m_df)\n",
    "\n",
    "\n",
    "def get_promotor_window(start, end, strand, meth_data=None):\n",
    "    if strand == '+':\n",
    "        tss = start\n",
    "    else:\n",
    "        tss = end\n",
    "    if not meth_data is None: # fancy method to get promoter\n",
    "        scan_region = (np.max(tss-1000, 0), tss + 1000)\n",
    "        return calculate_promoter_window(scan_region[0], scan_region[1], meth_data)\n",
    "    else:\n",
    "        return np.max(tss-250, 0), tss + 250\n",
    "\n",
    "\n",
    "def calculate_promoter_window(scan_start, scan_end, meth_data, shift=50, size=200):\n",
    "    best_mean_meth = None\n",
    "    best_window = (None, None)\n",
    "    for i in range(int(scan_start), int(scan_end-size+1), shift):\n",
    "        # get mean meth for window\n",
    "        m_sites_window = meth_data[meth_data.Start.between(i, i+size)]\n",
    "        m_mean_window = m_sites_window.Beta_value_mean.mean()\n",
    "        \n",
    "        # decide on best window\n",
    "        if best_mean_meth is None: # first window\n",
    "            best_mean_meth = m_mean_window\n",
    "            best_window = i, i+size\n",
    "        else: # any of the later windows\n",
    "            if abs(best_mean_meth-m_mean_window) > 0.25: # large change in mean detected\n",
    "                break\n",
    "            else: # make the current window the best one\n",
    "                best_mean_meth = m_mean_window\n",
    "                best_window = i, i+size\n",
    "    return best_window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "MmhPt3lW3sQn"
   },
   "outputs": [],
   "source": [
    "def get_filenames(meth_raw_dir):\n",
    "    \"\"\"Extract all valid methylation files in dir.\n",
    "    \"\"\"\n",
    "    all_filenames = []\n",
    "    skipped_files = 0\n",
    "    for dname in os.listdir(meth_raw_dir):\n",
    "        sub_dirname = os.path.join(meth_raw_dir, dname)\n",
    "        if os.path.isdir(sub_dirname):\n",
    "            for fname in os.listdir(sub_dirname):\n",
    "                if fname.endswith('gdc_hg38.txt'):\n",
    "                    if fname.split('.')[2].endswith('450'): \n",
    "                        all_filenames.append(os.path.join(sub_dirname, fname))\n",
    "                    else:\n",
    "                       skipped_files += 1\n",
    "    print (\"Skipped {} 27k arrays\".format(skipped_files))\n",
    "    return all_filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "x14qzAW-3yvs"
   },
   "outputs": [],
   "source": [
    "def get_mean_betaval_for_sample(annotation_genes, methylation_levels):\n",
    "    \"\"\"Extract mean methylation levels in promoters/gene bodies for each annotated gene.\n",
    "    \n",
    "    This function looks for all the methylation sites that fall\n",
    "    inside all the promoters and gene bodies in the annotation genes df.\n",
    "    The annotation_genes df is expected to contain the columns\n",
    "    'promoter_start' and 'promoter_end'.\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    annotation_genes:           A dataframe containing a gene per row.\n",
    "                                It needs to have columns for promoter_start\n",
    "                                and promoter_end.\n",
    "    methylation_levels:         A dataframe that contains cpg sites per row.\n",
    "                                The sites are expected to be one bp long and\n",
    "                                should have the columns: 'Start', 'End' and \n",
    "                                Beta_value'.\n",
    "    Returns:\n",
    "    Two lists with the mean beta values for each promoter and the number of cpg sites\n",
    "    that support the promoter. The first list contains the mean beta values and\n",
    "    the second list contains the support.\n",
    "    \"\"\"\n",
    "    beta_vals_prom = []\n",
    "    beta_vals_gene = []\n",
    "    n_supports_prom = []\n",
    "    n_supports_gene = []\n",
    "    count = 0\n",
    "    for _, row in annotation_genes.iterrows():\n",
    "        # restrict possible sites to same chr\n",
    "        cpgs_on_chr = methylation_levels[methylation_levels.Chromosome == row.chr]\n",
    "        # promoter\n",
    "        m_sites_in_promoter = cpgs_on_chr[cpgs_on_chr.Start.between(row.promoter_start, row.promoter_end)]\n",
    "        beta_vals_prom.append(m_sites_in_promoter.Beta_value.mean())\n",
    "        n_supports_prom.append(m_sites_in_promoter.shape[0])\n",
    "        # gene body. Don't start with the start and end coordinates but rather with the end of the promoter\n",
    "        if row.strand == '+':\n",
    "            m_sites_gene = cpgs_on_chr[cpgs_on_chr.Start.between(row.promoter_end, row.end)]\n",
    "        else:\n",
    "            m_sites_gene = cpgs_on_chr[cpgs_on_chr.Start.between(row.start, row.promoter_start)]\n",
    "        beta_vals_gene.append(m_sites_gene.Beta_value.mean())\n",
    "        n_supports_gene.append(m_sites_gene.shape[0])\n",
    "        count += 1\n",
    "    return beta_vals_prom, beta_vals_gene, n_supports_prom, n_supports_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "f8g3w3Da32hD"
   },
   "outputs": [],
   "source": [
    "def get_float(string):\n",
    "    \"\"\" Convert a string into a float.\n",
    "    This function converts a string to float if the string represents a floating\n",
    "    point number (or integer). If that is not the case, it returns NaN.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    string:                     The string that is to be converted to float.\n",
    "    Returns:\n",
    "    A floating point number. Either with the value from `string` or numpy.nan\n",
    "    \"\"\"\n",
    "    try:\n",
    "        return float(string)\n",
    "    except:\n",
    "        return np.inf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "SBB9O-AD37VY"
   },
   "outputs": [],
   "source": [
    "def get_closest_gene(row):\n",
    "    \"\"\" Get the closest gene from a row of TCGA level 3 methylation data.\n",
    "    Extracts the gene name of the closest transcript promoter from\n",
    "    a row of TCGA DNA methylation data (level 3). Each of the lines\n",
    "    contain the coordinates of the CpG site together with information\n",
    "    on the closest gene around it. There are three relevant fields (columns)\n",
    "    for that: Gene_Symbol, Transcript_ID and Position_to_TSS.\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    row:                        A row from a TCGA DNA methylation Dataframe.\n",
    "                                Should contain the columns `Position_to_TSS`\n",
    "                                and `Gene_Symbol`.\n",
    "    \"\"\"\n",
    "    # extract rows\n",
    "    genes = np.array(row.Gene_Symbol.split(';'))\n",
    "    protein_coding_genes = np.array([i == 'protein_coding' for i in row.Gene_Type.split(';')])\n",
    "    dists = np.array([get_float(i) for i in row.Position_to_TSS.split(';')])\n",
    "\n",
    "    # remove non-protein-coding genes\n",
    "    dists = dists[protein_coding_genes]\n",
    "    genes = genes[protein_coding_genes]\n",
    "\n",
    "    # return closest gene and distance\n",
    "    if dists.shape[0] > 0:\n",
    "        idx = np.argmin(np.abs(dists))\n",
    "        return genes[idx], dists[idx]\n",
    "    else:\n",
    "        return None, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "2pRqPdrA3_uO"
   },
   "outputs": [],
   "source": [
    "def get_closest_5prime_transcript(row, gene_transcript_mapping):\n",
    "    \"\"\" Get the closest 5' transcript from a row of TCGA level 3 methylation data.\n",
    "    Extracts the transcript ID of the transcript whose TSS is closest to the\n",
    "    CpG site in question. This is computed from a row of TCGA DNA methylation\n",
    "    data (level 3). The function uses a pre-computed mapping from genes\n",
    "    to 5' transcripts for its calculation.\n",
    "    It first extracts the transcripts and distances to TSS for the CpG site\n",
    "    and then discards all that are not in the list of closest transcripts in\n",
    "    the mapping. Finally, from the valid transcripts the one with the closest\n",
    "    TSS is chosen and returned.\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    row:                        A row from a TCGA DNA methylation Dataframe.\n",
    "                                Should contain the columns `Position_to_TSS`\n",
    "                                and `Gene_Symbol`.\n",
    "    \n",
    "    Returns:\n",
    "    The transcript ID and distance to the transcript TSS of a CpG site.\n",
    "    This can be used as part of an apply call on a dataframe.\n",
    "    \"\"\"\n",
    "    # extract rows\n",
    "    transcripts = pd.Series(row.Transcript_ID.split(';'))\n",
    "    transcripts = transcripts.str.split('.').str[0] # remove transcript version number\n",
    "    \n",
    "    valid_transcripts = transcripts.isin(gene_transcript_mapping)\n",
    "    dists = np.array([get_float(i) for i in row.Position_to_TSS.split(';')])\n",
    "\n",
    "    # remove non-protein-coding genes\n",
    "    dists = dists[valid_transcripts]\n",
    "    transcripts = transcripts[valid_transcripts]\n",
    "\n",
    "    if dists.shape[0] > 0:\n",
    "        idx = np.argmin(np.abs(dists))\n",
    "        return np.array(transcripts)[idx], dists[idx]\n",
    "    else:\n",
    "        return None, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "dK8jSlZI4G-H"
   },
   "outputs": [],
   "source": [
    "def get_promoter_betaval_tcgaannotation(methylation_levels):\n",
    "    # write to DF which is the closest gene\n",
    "    x = methylation_levels.apply(get_closest_gene, axis=1)\n",
    "    methylation_levels['closest_gene'] = [i[0] for i in x]\n",
    "    methylation_levels['dist_closest_gene'] = [i[1] for i in x]\n",
    "\n",
    "    # filter\n",
    "    close_cpgs = methylation_levels[methylation_levels.dist_closest_gene.abs() < 1000]\n",
    "\n",
    "    # group and return\n",
    "    cpgs_for_genes = close_cpgs.groupby('closest_gene')\n",
    "    beta_values_prom = cpgs_for_genes.Beta_value.mean()\n",
    "    n_supports_prom = cpgs_for_genes.Beta_value.count()\n",
    "\n",
    "    return beta_values_prom, n_supports_prom\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "WbHbLOI74y81"
   },
   "outputs": [],
   "source": [
    "def get_gene2transcript_map(annotation_df):\n",
    "    \"\"\"Calculate a mapping from a gene to its most 5' transcript.\n",
    "    This function calculates a mapping from a gene to its most 5' transcript using a\n",
    "    GENCODE annotation file. For each transcript, the corresponding parent gene is\n",
    "    extracted together with its distance to the chromosome start. For genes on the\n",
    "    - strand, negative values are used.\n",
    "    Using a map/reduce pattern, the transcripts are grouped by parent genes and the\n",
    "    most 5' transcript per gene is chosen.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    annotation_df:              The GENCODE annotation file as pandas DataFrame.\n",
    "                                Usually, the basic annotation is preferred over the\n",
    "                                comprehensive.\n",
    "    Returns:\n",
    "    A DataFrame containing the gene symbol as index and the corresponding 5' transcript\n",
    "    as column.\n",
    "    \"\"\"\n",
    "    # get a mapping from gene to transcript\n",
    "    transcripts = annotation_df[annotation_df.type == 'transcript'].copy()\n",
    "    # extract all attributes of transcripts\n",
    "    transcript_attrs = []\n",
    "    for attrs in transcripts.attr.str.split(';'):\n",
    "        transcript_attrs.append({i.split('=')[0]:i.split('=')[1] for i in attrs})\n",
    "\n",
    "    # set the attributes to additional columns (remove transcript version number)\n",
    "    transcripts['ID'] = [i['ID'].strip().split('.')[0] for i in transcript_attrs]\n",
    "    transcripts['Parent'] = [i['Parent'] for i in transcript_attrs]\n",
    "    transcripts['Gene_Symbol'] = [i['gene_name'] for i in transcript_attrs]\n",
    "\n",
    "    # build column with the distance to the chromosome start (for transcripts on the - strand, take negative values)\n",
    "    transcripts['dist_to_chr_start'] = transcripts.start\n",
    "    transcripts.loc[transcripts.strand == '-', 'dist_to_chr_start'] = -transcripts.end\n",
    "    transcripts.set_index('ID', inplace=True)\n",
    "    # mapping is now simply the transcript in a group of a gene that has minimum distance\n",
    "    gene_transcript_mapping = transcripts.groupby('Gene_Symbol').dist_to_chr_start.idxmin()\n",
    "\n",
    "    return gene_transcript_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "SBooaccnEJAR"
   },
   "outputs": [],
   "source": [
    "def get_cpg_transcript_map(annotation_df, methylation_levels):\n",
    "    \"\"\"Calculates a mapping from CpG site to closest transcript.\n",
    "    \n",
    "    This function calculates a mapping from CpG site to gene promoter and\n",
    "    the distance to it. It makes use of the annotations in a TCGA methylation\n",
    "    file (level 3). For each CpG site, its closest transcripts and their distance\n",
    "    to the TSS are extracted. Then, all transcripts that are not the most 5'\n",
    "    transcripts of a gene are discarded and from the remaining transcripts, the\n",
    "    closest one is chosen. This implies that a CpG site can only belong to one\n",
    "    gene promoter.\n",
    "    The calculated map of CpG sites to transcript promoters is finally joined\n",
    "    with a mapping from transcripts to genes (computed from GENCODE annotation)\n",
    "    and the final mapping is returned.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    annotation_df:              The GENCODE annotation file as pandas DataFrame.\n",
    "                                Usually, the basic annotation is preferred over the\n",
    "                                comprehensive.\n",
    "    methylation_levels:         A TCGA level 3 methylation file. This will\n",
    "                                serve as template for all other samples, so\n",
    "                                if you use 450k data, this file should contain\n",
    "                                all the CpG sites you are interested in.\n",
    "    Returns:\n",
    "    A Dataframe with CpG sites as rows (Composite Element REF) and the gene/transcript\n",
    "    as columns. The columns contain the distance to the closest transcript of that CpG\n",
    "    site, the Ensembl ID of that transcript and the gene name of the transcript (Symbol).\n",
    "    \"\"\"\n",
    "    gene_transcript_map = get_gene2transcript_map(annotation_df)\n",
    "    # get the closest 5' transcript according to the gene transcript map\n",
    "    x = methylation_levels.apply(get_closest_5prime_transcript,\n",
    "                                 axis=1,\n",
    "                                 gene_transcript_mapping=gene_transcript_map)\n",
    "    methylation_levels['closest_transcript'] = [i[0] for i in x]\n",
    "    methylation_levels['dist_closest_transcript'] = [i[1] for i in x]\n",
    "    \n",
    "    # get transcript ID and Gene Symbol in mapping as new columns\n",
    "    gene_transcript_map_df = pd.DataFrame(gene_transcript_map)\n",
    "    gene_transcript_map_df.columns = ['Transcript_ID']\n",
    "    gene_transcript_map_df['Symbol'] = gene_transcript_map_df.index\n",
    "\n",
    "    # construct the mapping\n",
    "    cpgs_with_genes = methylation_levels[~methylation_levels.closest_transcript.isnull()]\n",
    "    cpgs_with_genes.drop(['Transcript_ID', 'Position_to_TSS'], axis=1, inplace=True)\n",
    "    cpg_gene_mapping = cpgs_with_genes.merge(gene_transcript_map_df,\n",
    "                                             left_on='closest_transcript',\n",
    "                                             right_on='Transcript_ID'\n",
    "                                            )\n",
    "    cpg_gene_mapping.set_index('Composite Element REF', inplace=True)\n",
    "    mapping_cols = ['dist_closest_transcript', 'Transcript_ID', 'Symbol']\n",
    "    return cpg_gene_mapping[mapping_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_meth_df_from_mapping(cpg_gene_map, path_name, clean=False):\n",
    "    \"\"\"Computes promoter DNA methylation from a CpG to gene mapping.\n",
    "    This function computes the average DNA methylation at promoters from\n",
    "    a mapping of CpG sites to genes. It essentially only joins the mapping\n",
    "    with the actual beta values from the sample in question.\n",
    "    It then aggregates the genes together, returning a DataFrame with\n",
    "    the average DNA methylation at the promoter of each gene and the\n",
    "    cancer type of the individual.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    cpg_gene_map:               A mapping between CpG sites and genes/transcripts.\n",
    "                                Can be computed from TCGA annotations to the methylation\n",
    "                                data. @see get_cpg_transcript_map\n",
    "    path_name:                  The path to the methylation file. Should be in TCGA format,\n",
    "                                thus containing information on close-by genes additionally\n",
    "                                to the pure coordinates and beta value.\n",
    "    clean:                      Whether to use potential previous runs and data written into\n",
    "                                the download directory or not. If changing the method of how\n",
    "                                DNA methylation is computed, clean=True is recommended.\n",
    "    Returns:\n",
    "    A dataframe with methylation levels at promoters and number of CpG sites per gene promoter as\n",
    "    columns and gene names as rows and the cancer type of the sample.\n",
    "    Ideally, there should be as many genes as there are in the mapping file.\n",
    "    The columns of the returned dataframe will have the format:\n",
    "    <sample_id>|<cancer_type>|<beta_val/support> (separator is |)\n",
    "    \"\"\"\n",
    "    sub_dirname = os.path.dirname(path_name)\n",
    "    fname = os.path.basename(path_name)\n",
    "    avg_meth_result_path = os.path.join(sub_dirname,\n",
    "            'avg_methylation_tcgaannot.tsv'\n",
    "            )\n",
    "    cancer_type = fname.split('.')[1].split('_')[1].strip().lower()\n",
    "    sample_id = fname.split('.')[-3].strip()\n",
    "\n",
    "    if not os.path.isfile(avg_meth_result_path) or clean:\n",
    "        meth_df = load_methylation_file(path_name)\n",
    "        meth_df.drop(['Transcript_ID', 'Gene_Symbol', 'Gene_Type', 'Position_to_TSS',\n",
    "                      'CGI_Coordinate', 'Feature_Type'],\n",
    "                      axis=1, inplace=True)\n",
    "        meth_df.set_index('Composite Element REF', inplace=True)\n",
    "        meth_annot = meth_df.join(cpg_gene_map)\n",
    "        meth_annot_inrange = meth_annot[meth_annot.dist_closest_transcript.abs() < 1000]\n",
    "        res = pd.DataFrame(meth_annot_inrange.dropna().groupby('Symbol').Beta_value.mean())\n",
    "        res['support_promoter'] = meth_annot_inrange.dropna().groupby('Symbol').Beta_value.count()\n",
    "        colnames = ['mean_beta_value_promoter', 'support_promoter']\n",
    "        cols = ['{}|{}|{}'.format(sample_id, cancer_type, i) for i in colnames]\n",
    "        res.columns = cols\n",
    "        res.to_csv(avg_meth_result_path, sep='\\t')\n",
    "    else:\n",
    "        res = pd.read_csv(avg_meth_result_path, sep='\\t')\n",
    "        res.set_index('Symbol', inplace=True)\n",
    "    return res, cancer_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "-ArfHAu659Iu"
   },
   "outputs": [],
   "source": [
    "def get_meth_df_for_sample(annotation_df, path_name, clean=False, tcga_annot=True):\n",
    "    \"\"\"Calculates the mean methylation DF for one sample.\n",
    "    \n",
    "    This function computes the mean methylation level at\n",
    "    the promoter and gene body for all genes in the annotation\n",
    "    dataframe using the methylation data specified in the path\n",
    "    name. Note that this must be in the TCGA methylation format.\n",
    "    The annotation dataframe is expected to contain the promoter\n",
    "    region in the columns 'promoter_start' and 'promoter_end'.\n",
    "    If the directory of the file given by 'path_name' contains\n",
    "    already a file 'avg_methylation.tsv', this will be loaded\n",
    "    and returned simply instead of re-extracting the methylation\n",
    "    values. If clean is set to True, the extraction takes place\n",
    "    in any case.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    annotation_df:              The annotation file as dataframe. Can be computed using\n",
    "                                @see load_annotation_gff from the gencode annotation.\n",
    "    path_name:                  The path to the methylation file. Should be in TCGA format,\n",
    "                                thus containing information on close-by genes additionally\n",
    "                                to the pure coordinates and beta value.\n",
    "    clean:                      Whether to use potential previous runs and data written into\n",
    "                                the download directory or not. If changing the method of how\n",
    "                                DNA methylation is computed, clean=True is recommended.\n",
    "    tcga_annot:                 Whether or not to use the annotation from TCGA to each\n",
    "                                individual CpG site. Otherwise it uses promoter windows\n",
    "                                around the TSS of a gene.\n",
    "    Returns:\n",
    "    A dataframe with methylation levels at promoters and number of CpG sites per gene promoter as\n",
    "    columns and gene names as rows and the cancer type of the sample.\n",
    "    Ideally, there should be as many genes as there are in the mapping file.\n",
    "    The columns of the returned dataframe will have the format:\n",
    "    <sample_id>|<cancer_type>|<beta_val/support> (separator is |)\n",
    "    \"\"\"\n",
    "    sub_dirname = os.path.dirname(path_name)\n",
    "    fname = os.path.basename(path_name)\n",
    "    if tcga_annot:\n",
    "        avg_meth_result_path = os.path.join(sub_dirname,\n",
    "                'avg_methylation_tcgaannot.tsv'\n",
    "                )\n",
    "    else:\n",
    "        avg_meth_result_path = os.path.join(sub_dirname,\n",
    "                        'avg_methylation.tsv'\n",
    "                        )\n",
    "    # get the cancer type first\n",
    "    cancer_type = fname.split('.')[1].split('_')[1].strip().lower()\n",
    "    sample_id = fname.split('.')[-3].strip()\n",
    "    if not os.path.isfile(avg_meth_result_path) or clean:\n",
    "        # get mean methylation levels around promoters\n",
    "        meth_df = load_methylation_file(path_name)\n",
    "        if tcga_annot:\n",
    "            beta_prom, sup_prom = get_promoter_betaval_tcgaannotation(meth_df)\n",
    "            res = pd.DataFrame([beta_prom, sup_prom]).T\n",
    "            colnames = ['mean_beta_value_promoter', 'support_promoter']\n",
    "            cols = ['{}|{}|{}'.format(sample_id, cancer_type, i) for i in colnames]\n",
    "            res.columns = cols\n",
    "        else:\n",
    "            beta_p, beta_g, sup_p, sup_g = get_mean_betaval_for_sample(annotation_df,\n",
    "                                                                       meth_df)\n",
    "            # construct dataframe from results\n",
    "            res = pd.DataFrame([beta_p, sup_p, beta_g, sup_g]).T\n",
    "            colnames = ['mean_beta_value_promoter', 'support_promoter',\n",
    "                        'mean_beta_value_gene', 'support_gene']\n",
    "            cols = ['{}|{}|{}'.format(sample_id, cancer_type, i) for i in colnames]\n",
    "            res.columns = cols\n",
    "            res.set_index(annotation_df.Symbol, inplace=True) # assumes same order\n",
    "\n",
    "        # write the average beta values (at promoters) to file\n",
    "        res.to_csv(avg_meth_result_path, sep='\\t')\n",
    "    else:\n",
    "        res = pd.read_csv(avg_meth_result_path, sep='\\t')\n",
    "        res.set_index('Symbol', inplace=True)\n",
    "    return res, cancer_type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---MAIN CODE FOR PREPROCESSING---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "OizHPELAi4uT"
   },
   "outputs": [],
   "source": [
    "relevant_genes = None\n",
    "annotation_df = load_annotation_gff('C:/Users/Shreya/Downloads/methluad/gencode.v39.annotation.gff3')\n",
    "print(annotation_df.head(3))\n",
    "print (\"Loaded Annotation file with {} genes\".format(annotation_df.shape[0]))\n",
    "a_df = pd.read_csv('C:/Users/Shreya/Downloads/methluad/gencode.v39.annotation.gff3', sep='\\t', skiprows=7,\n",
    "                        header=None,\n",
    "                        names=['chr', 'source', 'type', 'start', 'end', 'score',\n",
    "                                'strand', 'phase', 'attr'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get file names of DNA methylation downloads\n",
    "all_files = get_filenames('C:\\Users\\Shreya\\Downloads\\methluad')\n",
    "print (\"Found {} methylation profiles\".format(len(all_files)))\n",
    "\n",
    "print (\"Computing CpG site to transcript map\")\n",
    "# use an EPIC 850k array meth file to calculate mapping as it has >2k probes\n",
    "# cpg_mapping = get_cpg_transcript_map(a_df, load_methylation_file(all_files[0]))\n",
    "# cpg_mapping.to_csv('C:/Users/Shreya/Downloads/methluad/cpg_genemap_file', sep='\\t')\n",
    "\n",
    "# if cpg genemap file is already saved, use this\n",
    "cpg_mapping = pd.read_csv(\"C:/Users/Shreya/Downloads/methluad/cpg_genemap_file.tsv\", sep='\\t')\n",
    "cpg_mapping.set_index('Composite Element REF', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print (\"Applying cpg 2 transcript mapping for all samples (takes time and uses processors)\")\n",
    "tumor_files= [f for f in all_files if f.split('\\\\')[-1].split('.')[5].split('-')[3].startswith('0')]\n",
    "normal_files= [f for f in all_files if f.split('\\\\')[-1].split('.')[5].split('-')[3].startswith('1')]\n",
    "\n",
    "print (len(tumor_files))\n",
    "print (len(normal_files))\n",
    "\n",
    "tumor_results = Parallel(n_jobs=1)(delayed(get_meth_df_from_mapping)(cpg_mapping,\n",
    "                                                                                 f,\n",
    "                                                                                 True) for f in tumor_files)\n",
    "normal_results = Parallel(n_jobs=1)(delayed(get_meth_df_from_mapping)(cpg_mapping,\n",
    "                                                                                 f,\n",
    "                                                                                 True) for f in normal_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tumor_samples_preprocessed = [i[0] for i in tumor_results]\n",
    "normal_samples_preprocessed = [i[0] for i in normal_results]\n",
    "\n",
    "# join samples to form one df with all samples as columns\n",
    "total_tumor_df = pd.concat(tumor_samples_preprocessed, axis=1)\n",
    "total_normal_df = pd.concat(normal_samples_preprocessed, axis=1)\n",
    "total_tumor_df.to_csv('C:/Users/Shreya/Downloads/methluad/mean_meth_tumor_matrix.tsv', sep='\\t')\n",
    "total_tumor_df.to_csv('C:/Users/Shreya/Downloads/methluad/mean_meth_normal_matrix.tsv', sep='\\t')\n",
    "print (\"Finished. Written tumor mean methylation matrix for {} samples to disk\".format(len(tumor_samples_preprocessed)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make it the same genes for both dataframes and verify\n",
    "common_genes = total_tumor_df.index.intersection(total_normal_df.index)\n",
    "total_tumor_df = total_tumor_df.reindex(common_genes)\n",
    "total_normal_df = total_normal_df.reindex(common_genes)\n",
    "assert ((total_normal_df.index == total_tumor_df.index).all())\n",
    "\n",
    "# remove support columns\n",
    "total_tumor_df = total_tumor_df.drop(total_tumor_df.filter(regex='support').columns, axis=1)\n",
    "total_normal_df = total_normal_df.drop(total_normal_df.filter(regex='support').columns, axis=1)\n",
    "\n",
    "# compute log2 fold changes (for each sample, divide by mean normal values. Then compute mean across samples)\n",
    "fc = np.log2(total_tumor_df.divide(total_normal_df.mean(axis=1), axis=0))\n",
    "fc_nan = fc.replace([np.inf, -np.inf], np.nan)\n",
    "print (\"Got {} invalid values after computing log2 fold changes\".format(fc_nan.isnull().sum().sum()))\n",
    "ratio = fc_nan.dropna(axis=0) # remove NaN and inf (from division by 0 or 0+eta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # calculate mean of all log2fc samples and save to directory\n",
    "final_mean_meth=ratio.mean(axis=1)\n",
    "print (final_mean_meth)\n",
    "final_mean_meth.to_csv('C:/Users/Shreya/Downloads/methluad/mean_meth_final_matrix.tsv', sep='\\t')\n",
    "\n",
    "# save hdf5 file to train gcn\n",
    "final_mean_meth=pd.read_csv('C:/Users/Shreya/Downloads/methluad/mean_meth_final_matrix.tsv', sep='\\t')\n",
    "final_mean_meth.to_hdf('C:/Users/Shreya/Downloads/methluad/meth_hdf5_matrix.h5','methylation_sample_matrix', complevel=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---USE THRESHOLD TO IDENTIFY HYPO AND HYPERMETHYLATED GENES---\n",
    "https://www.frontiersin.org/articles/10.3389/fgene.2021.717557/full\n",
    "https://wjso.biomedcentral.com/articles/10.1186/s12957-021-02124-6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0        0.022220\n",
      "1       -0.155013\n",
      "2       -0.088432\n",
      "3       -0.203156\n",
      "4       -0.136404\n",
      "           ...   \n",
      "25311   -0.049543\n",
      "25312    0.076625\n",
      "25313   -0.087765\n",
      "25314   -0.285657\n",
      "25315   -0.086860\n",
      "Name: betaval, Length: 25316, dtype: float64\n",
      "          gene   betaval\n",
      "114     ACADVL  0.519990\n",
      "138     ACOT12  0.554892\n",
      "172      ACTA1  0.809890\n",
      "220     ACVRL1  0.800762\n",
      "229     ADAM11  0.597461\n",
      "...        ...       ...\n",
      "25241    ZNF90  0.537971\n",
      "25264      ZP4 -0.501033\n",
      "25275   ZSCAN1  1.099927\n",
      "25284  ZSCAN23  1.019288\n",
      "25289  ZSCAN5A  0.853233\n",
      "\n",
      "[2088 rows x 2 columns]\n",
      "              gene   betaval\n",
      "172          ACTA1  0.809890\n",
      "220         ACVRL1  0.800762\n",
      "237         ADAM23  0.902087\n",
      "253       ADAMTS16  1.315819\n",
      "254    ADAMTS16-DT  1.125083\n",
      "...            ...       ...\n",
      "25237       ZNF880  0.918941\n",
      "25238       ZNF888  0.986900\n",
      "25275       ZSCAN1  1.099927\n",
      "25284      ZSCAN23  1.019288\n",
      "25289      ZSCAN5A  0.853233\n",
      "\n",
      "[1427 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# log2fc threshold to filter most methylated genes\n",
    "meth=pd.read_csv('C:/Users/Shreya/Downloads/methluad/mean_meth_final_matrix.tsv', sep='\\t')\n",
    "meth.columns=['gene','betaval']\n",
    "print (meth['betaval'])\n",
    "DMG5=meth.loc[meth['betaval'].abs()> 0.5]\n",
    "print (DMG5)\n",
    "DMG5.to_csv('C:/Users/Shreya/Downloads/methluad/DMG5.tsv', sep='\\t')\n",
    "\n",
    "DMG6=meth.loc[meth['betaval'].abs()> 0.6]\n",
    "DMG6.to_csv('C:/Users/Shreya/Downloads/methluad/DMG6.tsv', sep='\\t')\n",
    "print (DMG6)\n",
    "\n",
    "# compare with beta val thresholds of tumor samples only\n",
    "# meth_tumor=pd.read_csv('C:/Users/Shreya/Downloads/methluad/mean_meth_tumor_matrix.tsv', sep='\\t')\n",
    "# meth_tumor = meth_tumor.drop(meth_tumor.filter(regex='support').columns, axis=1)\n",
    "# meth_tumor=meth_tumor.mean(axis=1)\n",
    "# # meth_tumor.columns=['betaval']\n",
    "# print (meth_tumor)\n",
    "# hyper=meth_tumor[meth_tumor> 0.8]\n",
    "# hypo=meth_tumor[meth_tumor< 0.2]\n",
    "# print (hyper)\n",
    "# print (hypo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---BUILD NETWORK AND METHYLATION CONTAINER FOR GCN---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "!pip install mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene\n",
    "# Symbol to entrez_id\n",
    "def get_entrez_from_symbol(list_of_symbols):\n",
    "    # get Ensembl IDs for gene names\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    res = mg.querymany(list_of_symbols,\n",
    "                       scopes='symbol',\n",
    "                       fields='entrezgene',\n",
    "                       species='human', returnall=True\n",
    "                      )\n",
    "\n",
    "    def get_symbol_and_entrez(x):\n",
    "        if 'entrezgene' in x:\n",
    "            entrez_id = x['entrezgene']\n",
    "            symbol = x['query']\n",
    "            return [symbol, entrez_id]\n",
    "        else:\n",
    "            return [x['query'], None]\n",
    "    output = [get_symbol_and_entrez(d) for d in res['out']]\n",
    "    # now, retrieve the names and IDs from a dictionary and put in DF\n",
    "    output = pd.DataFrame(output, columns=['Symbol', 'Entrez_ID'])\n",
    "    output.dropna(axis=0, inplace=True)\n",
    "    return output\n",
    "\n",
    "def get_symbols_from_ensembl(list_of_ensembl_ids):\n",
    "    # get Ensembl IDs for gene names\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    res = mg.querymany(list_of_ensembl_ids,\n",
    "                       scopes='ensembl.gene',\n",
    "                       fields='symbol',\n",
    "                       species='human', returnall=True\n",
    "                      )\n",
    "\n",
    "    def get_symbol_and_ensembl(d):\n",
    "        if 'symbol' in d:\n",
    "            return [d['query'], d['symbol']]\n",
    "        else:\n",
    "            return [d['query'], None]\n",
    "\n",
    "    node_names = [get_symbol_and_ensembl(d) for d in res['out']]\n",
    "    # now, retrieve the names and IDs from a dictionary and put in DF\n",
    "    node_names = pd.DataFrame(node_names, columns=['Ensembl_ID', 'Symbol'])\n",
    "    node_names.dropna(axis=0, inplace=True)\n",
    "    return node_names\n",
    "\n",
    "def get_ensembl_from_symbol(list_of_gene_symbols):\n",
    "    # get Ensembl IDs for gene names\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    res = mg.querymany(list_of_gene_symbols,\n",
    "                       scopes='symbol, refseq, uniprot',\n",
    "                       fields='ensembl.gene',\n",
    "                       species='human', returnall=True\n",
    "                      )\n",
    "\n",
    "    # now, retrieve the names and IDs from a dictionary and put in DF\n",
    "    def get_name_and_id(x):\n",
    "        if 'ensembl' in x:\n",
    "            ens_id = x['ensembl'][0]['gene'] if type(x['ensembl']) is list else x['ensembl']['gene']\n",
    "            symbol = x['query']\n",
    "            return [symbol, ens_id]\n",
    "        else:\n",
    "            return [x['query'], None]\n",
    "\n",
    "    ens_ids = [get_name_and_id(x) for x in res['out']]\n",
    "    \n",
    "    node_names = pd.DataFrame(ens_ids, columns=['Symbol', 'Ensembl_ID'])\n",
    "    node_names.dropna(axis=0, inplace=True)\n",
    "    #node_names.drop_duplicates(inplace=True)\n",
    "    return node_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load Methylation data\n",
    "meth_idc= pd.read_csv('/content/drive/MyDrive/DL_Genomics_Paper/Dataset_GDC/Methylation/Meth_IDC.csv')\n",
    "meth_ilc= pd.read_csv('/content/drive/MyDrive/DL_Genomics_Paper/Dataset_GDC/Methylation/Meth_ILC.csv')\n",
    "meth_idc.rename(columns = {'0': 'Mean'}, inplace = True)\n",
    "meth_ilc.rename(columns = {'0': 'Mean'}, inplace = True)\n",
    "meth_idc.set_index('Symbol', inplace=True)\n",
    "meth_ilc.set_index('Symbol', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load ppi network\n",
    "import networkx as nx\n",
    "ppi_network = pd.read_csv('C:\\Users\\Shreya\\Downloads\\methluad\\pcnet_edgelist.tsv.gz', sep='\\t', compression='gzip')\n",
    "ppi_graph = nx.from_pandas_edgelist(df=ppi_network, source='Source_Name', target='Target_Name')\n",
    "ppi_network = nx.to_pandas_adjacency(G=ppi_graph)\n",
    "print (\"Network has {} nodes/genes\".format(ppi_network.shape[0]))\n",
    "#get ensembl_id of ppi_nodes\n",
    "nodes=get_ensembl_from_symbol(ppi_network.index)\n",
    "# Delets nodes in PPI that don't have an ensemble ID\n",
    "node_list=nodes.Symbol\n",
    "ppi_network=ppi_network.loc[node_list]\n",
    "ppi_network=ppi_network.loc[:,node_list]"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "name": "meth_preprocess_emogi.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
